File: src/plugins/task/lib/content.ts
import type { AnyTask, TaskContent, TaskResult } from '../core/types'

/** Проверка: результат по задаче + факт прохождения (неважно, верно или нет). */
export function withResult(content: TaskContent, taskId: string, result: TaskResult): TaskContent {
  return {
    ...content,
    results: { ...content.results, [taskId]: result },
    completed: { ...content.completed, [taskId]: true },
  }
}

/**
 * «Пройти заново»: ответ и результат стираются — задача снова в состоянии
 * «не решалась»; факт прохождения остаётся.
 */
export function withRestart(content: TaskContent, taskId: string): TaskContent {
  const answers = { ...content.answers }
  const results = { ...content.results }
  delete answers[taskId]
  delete results[taskId]

  return { ...content, answers, results }
}

/**
 * Правка содержания задачи: ответ, результат и факт прохождения сбрасываются —
 * задача изменилась и считается непройденной до новой проверки.
 */
export function withTaskEdited(content: TaskContent, taskId: string, next: AnyTask): TaskContent {
  const answers = { ...content.answers }
  const results = { ...content.results }
  const completed = { ...content.completed }
  delete answers[taskId]
  delete results[taskId]
  delete completed[taskId]

  return {
    ...content,
    tasks: content.tasks.map((task) => (task.id === taskId ? next : task)),
    answers,
    results,
    completed,
  }
}

/** Удаление задачи: убирается вместе со своим ответом, результатом и флагом прохождения. */
export function withTaskDeleted(content: TaskContent, taskId: string): TaskContent {
  const answers = { ...content.answers }
  const results = { ...content.results }
  const completed = { ...content.completed }
  delete answers[taskId]
  delete results[taskId]
  delete completed[taskId]

  return {
    ...content,
    tasks: content.tasks.filter((task) => task.id !== taskId),
    answers,
    results,
    completed,
  }
}

File: src-tauri/src/plugins/task/service/data/content.rs
use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

use super::answers::TaskAnswerData;
use super::difficulty::CustomDifficultyData;
use super::tasks::TaskData;

/// Результат проверки задачи: `'correct' | 'incorrect'` на проводе и в БД.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum TaskResultData {
    Correct,
    Incorrect,
}

impl TaskResultData {
    /// Wire/DB-значение результата.
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Correct => "correct",
            Self::Incorrect => "incorrect",
        }
    }

    /// Разбор значения из БД (CHECK-констрейнт гарантирует корректность).
    pub fn parse(value: &str) -> Option<Self> {
        match value {
            "correct" => Some(Self::Correct),
            "incorrect" => Some(Self::Incorrect),
            _ => None,
        }
    }
}

/// Попытка прохождения задачи — снимок ответа на момент проверки.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TaskAttemptData {
    pub id: String,
    pub task_id: String,
    pub seq: i64,
    pub answer: Option<TaskAnswerData>,
    pub result: TaskResultData,
    pub checked_at: i64,
}

/// Агрегат контента task-ресурса.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TaskContentData {
    pub tasks: Vec<TaskData>,
    pub difficulties: Vec<CustomDifficultyData>,
    pub answers: BTreeMap<String, TaskAnswerData>,
    pub results: BTreeMap<String, TaskResultData>,
    pub completed: BTreeMap<String, bool>,
}

/// Снапшот контента ресурса вместе с таймстампами корня.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TaskSnapshotData {
    pub resource_id: String,
    pub content: TaskContentData,
    pub created_at: i64,
    pub updated_at: i64,
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn result_uses_lowercase_wire_names() {
        assert_eq!(
            serde_json::to_value(TaskResultData::Correct).unwrap(),
            json!("correct")
        );
        assert_eq!(
            serde_json::to_value(TaskResultData::Incorrect).unwrap(),
            json!("incorrect")
        );
        assert_eq!(
            TaskResultData::parse("correct"),
            Some(TaskResultData::Correct)
        );
        assert_eq!(TaskResultData::parse("wrong"), None);
    }
}
